datamigration/handler_test.go

```diff
@@ -0,0 +1,155 @@
+package datamigration
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/app"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/migration"
+	"github.com/iov-one/weave/store"
+	"github.com/iov-one/weave/weavetest"
+)
+
+func TestHandler(t *testing.T) {
+	var aliceCond = weavetest.NewCondition()
+
+	cases := map[string]struct {
+		Conds   []weave.Condition
+		Tx      weave.Tx
+		WantErr *errors.Error
+	}{
+		"successful migration": {
+			Conds: []weave.Condition{aliceCond},
+			Tx: &weavetest.Tx{
+				Msg: &ExecuteMigrationMsg{
+					Metadata:    &weave.Metadata{Schema: 1},
+					MigrationID: "first migration",
+				},
+			},
+			WantErr: nil,
+		},
+		"missing auth": {
+			Conds: []weave.Condition{},
+			Tx: &weavetest.Tx{
+				Msg: &ExecuteMigrationMsg{
+					Metadata:    &weave.Metadata{Schema: 1},
+					MigrationID: "first migration",
+				},
+			},
+			WantErr: errors.ErrUnauthorized,
+		},
+		"missing auth and wrong chain": {
+			Conds: []weave.Condition{},
+			Tx: &weavetest.Tx{
+				Msg: &ExecuteMigrationMsg{
+					Metadata:    &weave.Metadata{Schema: 1},
+					MigrationID: "second migration",
+				},
+			},
+			WantErr: errors.ErrUnauthorized,
+		},
+		"wrong chain migration": {
+			Conds: []weave.Condition{aliceCond},
+			Tx: &weavetest.Tx{
+				Msg: &ExecuteMigrationMsg{
+					Metadata:    &weave.Metadata{Schema: 1},
+					MigrationID: "second migration",
+				},
+			},
+			WantErr: errors.ErrChain,
+		},
+		"executed migration cannot be executed again": {
+			Conds: []weave.Condition{aliceCond},
+			Tx: &weavetest.Tx{
+				Msg: &ExecuteMigrationMsg{
+					Metadata:    &weave.Metadata{Schema: 1},
+					MigrationID: "zero migration",
+				},
+			},
+			WantErr: errors.ErrState,
+		},
+	}
+
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			defer withNewRegister()()
+
+			db := store.MemStore()
+			migration.MustInitPkg(db, "datamigration")
+
+			MustRegister(
+				"zero migration",
+				Migration{
+					ChainID:         "testchain",
+					RequiredSigners: []weave.Address{aliceCond.Address()},
+					Migrate:         func(context.Context, weave.KVStore) error { panic("never to be called") },
+				},
+			)
+
+			MustRegister(
+				"first migration",
+				Migration{
+					ChainID:         "testchain",
+					RequiredSigners: []weave.Address{aliceCond.Address()},
+					Migrate:         func(context.Context, weave.KVStore) error { return nil },
+				},
+			)
+			MustRegister(
+				"second migration",
+				Migration{
+					ChainID:         "staging",
+					RequiredSigners: []weave.Address{aliceCond.Address()},
+					Migrate:         func(context.Context, weave.KVStore) error { return nil },
```

Comment:
Can you write a mock migration function test that actually migrates data to demonstrate how to write functions?

Reply:
I think this is outside of the scope of this functionality. I give you the database and it is up to the user to decide what to do. You will probably use a bucket to create/update/delete entities.